### ISSUE TITLE
Trying to connect from process to Tensorflow Server, cannot use frozen graph

### ISSUE THREAD

**Issue by hcl14** [tensorflow/tensorflow#23147]
Hi everyone, I have an issue which can possibly be a bug, for which I am trying to find solution for quite some time:

Stack Overflow:
https://stackoverflow.com/questions/52700621/tensorflow-server-i-dont-want-to-initialize-global-variables-for-every-session

Discuss:
https://groups.google.com/a/tensorflow.org/forum/#!topic/discuss/rARPP1_ilNA

My story in short is that I create processes dynamically in Python and do model predictions using a session for Tensorflow Server `tf.train.Server`. Graph is saved in global scope, as I hope that default python 'fork' mode will not create any overhead as long as data in RAM is not changed ("copy-on-change" policy). Anyway, it's too long and expensive for me to reload model for each new process. Perhaps there might be a more efficient way to define model on server side, but I don't know it.

The problem is that somewhy I need to run initialization of global variables (`sess.run(tf.global_variables_initializer())`) each time I create new process and open new session. I tried to do dummy sun of the model before locking graph so that all the variables must be initialized, but no success - I still get the uninitialized variable error. Can anyone help me with that?

I have a small reproducible example of the problem compatible with Tensorflow 1.11:

https://github.com/hcl14/Tensorflow-server-launched-from-child-process

The places in the code are marked with `"PROBLEM"` comment.

Also, maybe I am doing everything wrong and there is a better way to quickly do model predictions from dynamically created processes?

**Comment by hcl14** [comment]
When I tried to call model this code from persistent process (i.e. not spawned dynamically for just one call, but created once on program start), I've got this error:

`Fetch argument <tf.Operation 'init' type=NoOp> cannot be interpreted as a Tensor. (Operation name: "init" op: "NoOp"  is not an element of this graph.)`

Following [stackoverflow post](https://stackoverflow.com/questions/47918382/explicit-tensorflow-session-gives-fetch-error-in-tensorflow-nmt), I changed in model inference:

`sess = tf.Session('grpc://'+tf_server_address, graph=graph)`

to 

`sess = tf.InteractiveSession('grpc://'+tf_server_address, graph=graph)`

I've got persistent process working and dynamically created process hang.

**Comment by tracihuang9** [comment]
@hcl14 Looks like your updated code snippet can fix the issue. 
You should be able to get frozen graphs for dynamic processes if you set up the clusters properly.

**Comment by wt-huang** [comment]
@hcl14 Looks like your updated code snippet can fix the issue.
You should be able to get frozen graphs for dynamic processes if you set up the clusters properly.

**Comment by hcl14** [comment]
@wt-huang Thanks for the answer, [I've implemented some of the suggestions](https://github.com/hcl14/Tensorflow-server-launched-from-child-process) I received on Stackoverflow for my question, which I believe became popular after I posted the issue here on Github. Those involve using eager execution and writing custom tensorflow http server. 
However, it's still unclear for me how to freeze graph for dynamic processes. Does it involve using `tf.train.Server` or just opening new default sessions for each process? Could you please point me to a good example?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @wt-huang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gadagashwini** [comment]
Hi @hcl14, This [tutorial](https://www.tensorflow.org/tfx/tutorials/serving/rest_simple) explains how to train and serve Tensorflow model with Tensorflow serving. Take a look at [this tutorial](https://www.tensorflow.org/tfx/serving/serving_basic) , which shows you how to use TensorFlow Serving components to export a trained TensorFlow model and use the standard `tensorflow_model_server` to serve it. Thank you!
